File: 2025/79. Word Search/Solution.java
class Solution {

    int[] dx = {0, 1, 0, -1};
    int[] dy = {1, 0, -1, 0};
    boolean[][] visited;

    public boolean exist(char[][] board, String word) {
        visited = new boolean[board.length][board[0].length];
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {  
                if (search(board, i, j, word, 0))
                    return true;          
            }
        }
        return false;
    }

    private boolean search(char[][] board, int i, int j, String word, int pos) {
        if (pos >= word.length() || word.charAt(pos) != board[i][j])
            return false;
        if (pos == word.length() - 1)
            return true;
        visited[i][j] = true;
        for (int k = 0; k < 4; k++) {
            int x = dx[k] + i;
            int y = dy[k] + j;
            if (!isOutOfBoundary(board, x, y)) {
                if (search(board, x, y, word, pos + 1))
                    return true;
            }
        }
        visited[i][j] = false;
        return false;
    }

    private boolean isOutOfBoundary(char[][] board, int i, int j) {
        if (i < 0 || j < 0 || i >= board.length || j >= board[0].length || visited[i][j])
            return true;
        return false;
    }
}

File: 2025/79. Word Search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        visited = [[False for _ in range(len(board[0]))] for _ in range(len(board))]

        def dfs(idx: int, i: int, j: int) -> bool:
            if idx >= len(word) or board[i][j] != word[idx]:
                return False
            if idx == len(word) - 1:
                return True
            visited[i][j] = True
            found = False
            for (x, y) in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                ii, jj = i + x, j + y
                if (ii >= 0 and jj >= 0 
                    and ii < len(board) and jj < len(board[0]) 
                    and not visited[ii][jj]):
                    found |= dfs(idx + 1, ii, jj)
            visited[i][j] = False
            return found

        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0] and dfs(0, i, j):
                    return True
        return False